Fail fast during flush of empty Records

anokfireball opened the pull request:
Do not write write records to disk unless scalar/non-scalar ambiguity is resolved. This is done when the user creates a record component that is either scalar/nonscalar and possibly constant.

Also fixes major bugs in the ADIOS1 backend concerning ```fileBased``` output.

Closes #287.

anokfireball committed:
Fail fast during flush of empty Records

Do not write write records to disk unless scalar/non-scalar ambiguity is
 resolved

anokfireball committed:
Adapt working code to fast failure

ADIOS1 backend not fully working yet

anokfireball commented on src/backend/MeshRecordComponent.cpp:
```diff
@@ -44,10 +44,16 @@ MeshRecordComponent::read()
     if( *aRead.dtype == DT::VEC_FLOAT )
         setPosition(a.get< std::vector< float > >());
+    else if( *aRead.dtype == DT::FLOAT )
+        setPosition(std::vector< float >(a.get< float >()));
```
This actually initializes a vector with
```cpp
static_cast< std::vector< float >::size_type >(a.get< float >())
```
default-initalized elements.
The intended behaviour is realized with
```cpp
setPosition(std::vector< float >({a.get< float >()}));
```

ax3l replied:
@C0nsultant do you want to update the branch with it?

anokfireball replied:
Will do before merging.

ax3l commented on src/backend/MeshRecordComponent.cpp:
```diff
@@ -44,10 +44,16 @@ MeshRecordComponent::read()
     else if( *aRead.dtype == DT::VEC_DOUBLE )
         setPosition(a.get< std::vector< double > >());
+    else if( *aRead.dtype == DT::DOUBLE )
+        setPosition(std::vector< double >(a.get< double >()));
```
same here

ax3l commented on src/backend/MeshRecordComponent.cpp:
```diff
@@ -44,10 +44,16 @@ MeshRecordComponent::read()
     else if( *aRead.dtype == DT::VEC_LONG_DOUBLE )
         setPosition(a.get< std::vector< long double > >());
+    else if( *aRead.dtype == DT::LONG_DOUBLE )
+        setPosition(std::vector< long double >(a.get< long double >()));
```
same here

ax3l commented on test/SerialIOTest.cpp:
```diff
@@ -1451,9 +1443,11 @@ TEST_CASE( "adios1_fileBased_write_test", "[serial][adios1]" )
     }
 
+    //TODO breaks openWriteFile logic when removed
+    //o.flush();
```
works now?

anokfireball replied:
No, unfortunately. I figured out where it all goes haywire, but not what causes the behavior.
It has to do with ADIOS's distinction between read/write files. The scenario under which I can reproduce this is when creating a ```fileBased``` Series and *not* flushing between creation of two Iterations (i.e. files). The problem then becomes that the file with the earliest creation time will have an on-disk size of 0Bytes and only the last file will be created as expected.
I doubt this is a problem with ADIOS itself, but the way the backend is using it:
File creation (i.e. opening a write file) is deferred until required (i.e. encountering a Dataset write operation). This might be a problem if no Dataset write occurs. Once this file is opened, it will not be closed until new defines or reads happen.
... and this is exactly where the problem happens. One of these calls happens at the wrong time (for ADIOS), so the file on disk will be empty. Unfortunately, ADIOS doesn't fail (only warns) when opening a 0Byte file (```adios_errno``` is not set, either). The ```SIGSEGV``` in the test happens when that faulty file is closed.

anokfireball replied:
(Likely) found the culprit:
In ```fileBased``` Series, *one* ADIOS group was used for *all* files. This works when only one file is open at a time, but that is not a constraint enforced in the ADIOS backends.
A solution is to create a group for every file handle. The commit alleviates the CI failure, but does not reliably use the group across the backend yet.

ax3l commented on examples/1_structure.cpp:
```diff
@@ -47,5 +47,10 @@ int main()
 
+    /* Required Records and RecordComponents are created automatically.
+     * Initialization has to be done explicitly by the user. */
+    electrons["position"][RecordComponent::SCALAR].resetDataset(dataset);
```
I just realized `resetDataset` in `(Patch)RecordComponent` lacks doxygen docs. What does it do, overwrite a dataset?

anokfireball replied:
It overwrites the default Dataset (single ```char```) with a user-defined dataset, including datatype, extent, chunking, compression, etc.
The Dataset is used when flushing a RecordComponent to disk, see [```RecordComponent::flush()```](https://github.com/openPMD/openPMD-api/blob/7501e3435913216812f3efe8707b30892b3aa915/src/RecordComponent.cpp#L97-L106) for example.

anokfireball replied:
At this point I am not sure if a default *valid* dataset is the best solution. It surely is helpful for prototyping, but having a ```flush()``` fail with an uninitialized, default dataset sends a clear message.

anokfireball committed:
Create ADIOS group for every file

ADIOS makes assumptions about the files in a group that do not hold in
our workflow. (Meta-)Data may differ between files in the same group.

anokfireball commented:
@ax3l OSX builds on Travis are failing after an upgrade to High Sierra. In particular, spack install of a compiler [fails](https://travis-ci.org/openPMD/openPMD-api/jobs/410981935#L949):
```
=> Error: No compilers found for operating system highsierra and target x86_64.
If previous installations have succeeded, the operating system may have been updated.
Compilers are defined for the following operating systems and targets:
	sierra-x86_64
	ubuntu14.04-x86_64
Run 'spack compiler find' to add compilers.
```
Can you look into that?

anokfireball committed:
Force write of at least one cell per dataset

ADIOS1 only creates a dataset on disk after a cell has been written

ax3l commented:
> OSX builds on Travis are failing after an upgrade to High Sierra

ugh, sorry about that. yes will update: #299

ref: https://blog.travis-ci.com/2018-07-19-xcode9-4-default-announce

anokfireball committed:
Unit tests for Record components without explicit dataset writes

anokfireball commented:
There is a pretty major blocker in the current ADIOS backend:
Attributes may only be defined *once* when using ```adios_define_attribute_byvalue()```. Otherwise, they will occur multiple times under multiple IDs.
The problem is that (borrowing from HDF5 behaviour), the backend was implemented under the assumption that successive definitions overwrite previous values.

```cpp
#define _NOMPI
#include <adios.h>


int main()
{
  adios_init_noxml(MPI_COMM_NULL);

  int64_t m_adios_group;
  int64_t m_adios_file;

  adios_declare_group(&m_adios_group, "att_test.bp", "", adios_stat_default);
  adios_select_method(m_adios_group, "POSIX", "", "");

  /* ############################################################################################################ */

  int i = 0;
  adios_define_attribute_byvalue(m_adios_group,
                                 "attribute", "/foo", adios_integer, 1, &i);

  i = 1;
  adios_define_attribute_byvalue(m_adios_group,
                                 "attribute", "/foo", adios_integer, 1, &i);

  /* ############################################################################################################ */

  adios_open(&m_adios_file, "att_test.bp", "att_test.bp", "w", MPI_COMM_NULL);
  adios_close(m_adios_file);

  adios_finalize(0);
  return 0;
}
```

leads to

```bash
 $ bpdump att_test.bp
[...]
 Attributes Count: 5
                Attribute Name (ID): attribute (1)
                Attribute Path: /foo
                Datatype: integer
                # of elements:   1
                Length in bytes: 4
                Value: 0

                Attribute Name (ID): attribute (2)
                Attribute Path: /foo
                Datatype: integer
                # of elements:   1
                Length in bytes: 4
                Value: 1
[...]
```

There are two options to explore for solving this problem:
- Deferring *all* attribute writes until the IOHandler is terminated. At that point we can determine which write happens last and discard all overwritten values.
- Using ```adios_define_attribute``` and write the attribute value to a variable. This is similar to what the backend already does for dynamic chunk sizes. Note ADIOS [doc](https://users.nccs.gov/~pnorbert/ADIOS-UsersManual-1.13.1.pdf):
> It is not appropriate to define a foating-point type attribute with high-precision, and cannot be used to define an array attribute. For those cases, see the next function below, ```adios_define_attribute_byvalue```.

ax3l commented:
The first approach looks more forward-thinking, e.g. for other backends that will have similar limitations.

anokfireball committed:
Defer Attribute write in ADIOS1

Do not flush Attribute value until termination of IOHandler. Variable
defines do not overwrite previous definitions. Only flush the very last
value as the definition.
Enable VERFIY macro for ADIOS backends (was disabled with #254).

anokfireball committed:
Avoid ADIOS1 overwriting files during READ_WRITE operations

anokfireball commented on src/IO/ADIOS/ADIOS1IOHandler.cpp:
```diff
@@ -48,32 +48,30 @@ ADIOS1IOHandlerImpl::ADIOS1IOHandlerImpl(AbstractIOHandler* handler)
+         * (happens when no Operation::WRITE_DATASET is performed) */
+        for( auto& f : m_filePaths )
+            if( m_openWriteFileHandles.find(f.second) == m_openWriteFileHandles.end() )
+                m_openWriteFileHandles[f.second] = open_write(f.first);
```
~For this to work with ```AccessType::READ_WRITE```, [```openFile()```](https://github.com/openPMD/openPMD-api/blob/2cc1cef79f9b75d181296956611dc6acd2aaed3f/src/IO/ADIOS/CommonADIOS1IOHandler.cpp#L453-L487) needs to set ```m_existsOnDisk``` so [```open_write()```](https://github.com/openPMD/openPMD-api/blob/2cc1cef79f9b75d181296956611dc6acd2aaed3f/src/IO/ADIOS/ADIOS1IOHandler.cpp#L247-L281) does not overwrite, but append.~
Done.

ax3l replied:
But does `openFile()` set this now?

anokfireball replied:
Yes it does, right [here](https://github.com/openPMD/openPMD-api/blob/414f2ddbdd1e0542d6f82aa32885e3adbfe7eef4/src/IO/ADIOS/CommonADIOS1IOHandler.cpp#L490).

anokfireball commented:
@ax3l Should be ready for review. I'll add a few more tests and a changelog for the recent PRs today.

anokfireball commented on src/IO/ADIOS/ADIOS1IOHandler.cpp:
```diff
@@ -43,38 +43,35 @@ namespace openPMD
+
+        for( auto& f : m_openWriteFileHandles )
+            close(f.second);
+    }
```
~This needs to be done in ```ParallelADIOS1IOHandler.cpp``` as well.~
Done.

ax3l replied:
done as far as I can see :)

anokfireball committed:
Unit Test for new fast failure

anokfireball committed:
Changelog for recent PRs

ax3l commented on include/openPMD/ParticleSpecies.hpp:
```diff
@@ -62,11 +62,11 @@ namespace traits
             auto& np = ret.particlePatches["numParticles"];
             auto& npc = np[RecordComponent::SCALAR];
-            npc.resetDataset(Dataset(Datatype::UINT64, {0}));
+            npc.resetDataset(Dataset(Datatype::UINT64, {1}));
```
Why do we set an extend of `1` here already? Doesn't this imply a first entry already? What if someone does not write data / writes exactly one entry later on? Just wondering about corner cases.

anokfireball replied:
ADIOS does not handle datasets without extent well. The default here is just used to enforce standard compliance and create deterministic behaviour across backends.
To address your concern about corner cases: [```PatchRecordComponent::resetDataset()```](https://github.com/openPMD/openPMD-api/blob/0.3.1-alpha/src/backend/PatchRecordComponent.cpp#L33-L42) should be used by users that do want to write actual data. It ~will~ should be replaced when writing production data.
If the user does *not* write dataset values ever, the created dataset will contain one potentially [default](https://github.com/openPMD/openPMD-api/blob/414f2ddbdd1e0542d6f82aa32885e3adbfe7eef4/src/RecordComponent.cpp#L107-L121)-initalized value.

ax3l commented on src/IO/ADIOS/ADIOS1IOHandler.cpp:
```diff
@@ -301,6 +293,19 @@ ADIOS1IOHandlerImpl::open_read( std::string const& name )
+{
+    int status;
+    int64_t group;
+    ADIOS_STATISTICS_FLAG noStatistics = adios_stat_no;
```
just a note: these statistics are quite interesting, but I never evaluated the RT costs in our use cases, so we turned them off by default.

if we could control from the outside we could evaluate them.

maybe good for an issue.

ax3l commented on src/RecordComponent.cpp:
```diff
@@ -103,6 +104,21 @@ RecordComponent::flush(std::string const& name)
+                if( m_chunks->empty() )
+                {
+                    /* Ensure at least one WRITE_DATASET per dataset occurs
+                     * ADIOS1 backend only creates a dataset after at least once cell has been written */
```
do we or ADIOS define the term cell?

anokfireball replied:
Interesting. ADIOS *does* use the same lingo for what I called cell. (See e.g. [```adios_define_mesh_unstructured```](https://users.nccs.gov/~pnorbert/ADIOS-UsersManual-1.13.1.pdf) on page 52).
I used it independently here, though.

ax3l commented on test/CoreTest.cpp:
```diff
@@ -102,8 +102,11 @@ TEST_CASE( "output_constructor_test", "[core]" )
 
-    o.iterations[1].meshes["foo"];
-    o.iterations[1].particles["bar"];
+    o.iterations[1].meshes["foo"]["baz"].resetDataset(Dataset(Datatype::DOUBLE, {1}));
```
can you add a (global) extend-zero write to the tests and it's defined behaviour? Basically `empty_record_test` but where the user sets the extend explicitly to `{0}`?

Do we already have a test that does a parallel write with extend > 0 (1D and ND) in which some of the MPI ranks contribute *locally* zero extends? This is quite often the case for e.g. particle data.

anokfireball replied:
You're probably misreading what is happening here.
```resetDataset()``` is not actually a data write-through to disk. It *defines* the *global* datatype, extent, compression (...) of each dataset. This is a collective call, every rank will have to call this with the same global data.
If the extent passed to ```resetDataset()``` is zero in any dimension (and is later flushed to disk), undefined behaviour in certain backends ensures.

Where you would actually pass the zero-extent in the scenario you're describing is in [```RecordComponent::storeChunk()```](https://github.com/openPMD/openPMD-api/blob/0.3.1-alpha/include/openPMD/RecordComponent.hpp#L167-L196), which triggers a ```Operation::WRITE_DATASET``` in the backends (see [HDF5](https://github.com/openPMD/openPMD-api/blob/0.3.1-alpha/src/IO/HDF5/HDF5IOHandler.cpp#L653-L735) and [ADIOS](https://github.com/openPMD/openPMD-api/blob/0.3.1-alpha/src/IO/ADIOS/CommonADIOS1IOHandler.cpp#L376-L416)).

#135 is for implementing such a local zero-extent write into a global dataset. Currently, there is no such test.

ax3l replied:
Ok, great. Just one thing:

> If the extent passed to resetDataset() is zero in any dimension (and is later flushed to disk), undefined behaviour in certain backends ensures.

Instead of undef behaviour, depending on backend, can we just implement defined similar behaviour for all backends? E.g., if global zero extent is given we throw (directly or in flush) or if zero is given we skip the write (and document that), etc.?

anokfireball replied:
> if global zero extent is given we throw (directly or in flush)

That's the behaviour I currently implement in #309. Skipping the undefined write (however verbose) does not appeal to me, as the dataset itself does not really make sense.

ax3l reviewed (commented):
Thank you so much, these are fantastic updates :sparkles: 

I have added some inline questions, probably most of them already addressed.

ax3l reviewed (approved):
Very nice! We can define a consistent behavior for global-zero extends in a follow-up.
